### PR TITLE
Dreamer

### PR DESCRIPTION
**Pull request by Rohan138**
## Types of changes
Still a WIP, but I've managed to add most of Dreamer. The main thing left is computing the loss in planning/dreamer_agent/DreamerAgent::train().

- [ Y] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ Y] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [ ] I have read the [**CONTRIBUTING**](../../CONTRIBUTING.md) document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by luisenp** [comment]
Hi @Rohan138, took a quick high-level glance and so far it looks good. I will start looking at the code in `algorithms/dreamer` and the `DreamerAgent` more closely now. 

I noticed you added some changes to other files like the pre-commit config, the requirements, pyproject.toml, etc. Were you running into some errors? If that's the case, would you mind opening a separate PR for these?

**Comment by luisenp** [comment]
@Rohan138 trying to pull from your fork as shown below, but I run into access permission errors. Could you see if I can get read access?

```bash
git checkout -b Rohan138-dreamer main
git pull git@github.com:Rohan138/mbrl-lib.git dreamer
```

**Comment by Rohan138** [comment]
I can add you to my fork, but do you want to add my repo as a remote and try `git checkout -t Rohan138/dreamer`?

**Comment by luisenp** [comment]
> git checkout -t Rohan138/dreamer

I tried a couple of things and I got errors:

image

image

**Comment by Rohan138** [comment]
Just sent you a contributor invite; maybe `https://github.com/Rohan138/mbrl-lib.git` would've worked better than `git@github`?

**Comment by luisenp** [comment]
HTTP worked for me before accepting the invitation, thanks!

**Comment by Rohan138** [comment]
On the non-Dreamer fixes:

- I renamed `pyproyect.toml` to `pyproject.toml`
- I removed the lines pinning `python==3.7` on all commits
- There's no black version 21.4b2 on pypi, so I upgraded to the latest stable version instead

I can definitely move these to a different PR, though.

**Comment by Rohan138** [comment]
I'm running performance tests on the five environments specified in `overrides/dreamer_*.yaml`, will report the results in another day or so.

Incidentally, it looks like the planet implementation in mbrl-lib (and hence dreamer) is significantly slower than that in the official [code](https://github.com/google-research/planet)(~10 hours) or other [pytorch](https://github.com/Kaixhin/PlaNet) (~20 hours) implementations on a V100; any ideas on why this might be?

I can look into this on my own, but thanks for your guidance and time in developing this library! :)

**Comment by luisenp** [comment]
@Rohan138 planning to spend most of today and then Friday playing around with your code. Is there anything in particular you'd like for me to focus on or help with? It seems I'm able to run Dreamer, but I haven't checked if it learns correctly yet. What's the current status?

**Comment by Rohan138** [comment]
Hi! So I tried running it, but despite the losses dropping, it doesn't seem to learn right now. [Here](https://drive.google.com/file/d/1AgeveNg2iDIBop26PJcFMFCOVvOGu_Ed/view?usp=sharing) are the results; I'm planning to look through the `agent.train()` implementation over the weekend to figure out what's going on.

**Review by luisenp — CHANGES_REQUESTED** [review]
Still looking into things, but left some initial comments. I'm also wondering about the way you are computing the value estimates in `_compute_return`, but need to look into this more carefully. Are you confident about that part of the code? I was thinking of maybe spending some time on Friday to write a utility function for this and maybe add some unit tests. Let me know if this would be useful.

**Comment by natolambert** [comment]
I'm thinking about starting a project that builds off the Dreamer style of dynamics model. If I spend a couple hours here and there on this PR, what would be most useful?

**Comment by Rohan138** [comment]
Sorry for the delay-I'll try to address all of the comments across the next day or so.

File: mbrl/algorithms/dreamer.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            cfg.overrides.batch_size,
+            0,  # no validation data
+            cfg.overrides.sequence_length,
+            max_batches_per_loop_train=cfg.overrides.num_grad_updates,
+            use_simple_sampler=True,
+        )
+        trainer.train(
+            dataset, num_epochs=1, batch_callback=model_batch_callback, evaluate=False
+        )
+        agent.train(dataset, num_epochs=1, batch_callback=agent_batch_callback)
```
[the remark this answers is not in the document]
Comment (Rohan138) [reply mbrl/algorithms/dreamer.py:171]: > I'm wondering if we should be passing a different iterator for training the agent. If I understand the paper correctly, the Dreamer agent is trained on trajectories whose start states are sampled from the experience buffer, but where all subsequent states are obtained by rolling out the model. In this case, we only need to sample individual transitions to get start states, and not full sequences, which is what `dataset` would return here.

I might have misunderstood the paper, but I'm not sure this is correct. In Algorithm 1 (Page 3), they:
1. Draw  `B` data sequences or episodes `{(a_t, o_t, r_t)} ~_{t=k}^{k+L}`. Here `k` is the outer variable looping over episodes, while `t` is the inner variable looping over timesteps in an episode.
2. Compute model states `s_t` for all `t` in `[k, k + L)` for all `k` in `B` using the RSSM transition model.
3. Imagine trajectories `{s_\tau, a_\tau}_{\tau = t}^{\tau = t + H}` from each state `s_t` in B, *not* just the initial state `s_k` in each episode.

I'm not sure if this explanation was clear, and I'll take another look at the prior implementations linked in the other comment to confirm.

[the remark this answers is not in the document]
Comment (luisenp) [reply mbrl/algorithms/dreamer.py:171]: Looking at the paper again, I think your interpretation is correct because the `Compute model states` step occurs for all sampled `o_t`, and then trajectories are imagined for all model states `s_t`. I find it a bit confusing how they are using the index `k`; I guess this increasing in increments of size `L`? That is, the j-th trajectory goes from `t=L*(j-1)` to `L*j - 1`? In any case, confirming with prior implementations is a good idea. 

Regarding the performance hit, one idea that wouldn't require a lot of changes would be to add get/set methods for random state of the iterator, so that we can have it return the same set of samples both for the model and agent loops. We should then be able to use the model trainer callback to store all computed model states, and pass them to the agent trainer in the correct order. 

Does that make sense?

**Comment by natolambert** [comment]
@luisenp @Rohan138  -- is there anything I or @RajGhugare19 can do to get this moving again?

**Comment by luisenp** [comment]
> @luisenp @Rohan138 -- is there anything I or @RajGhugare19 can do to get this moving again?

Hi @natolambert. Unfortunately, it's almost impossible for me at this point to take the lead in development, due to other more pressing commitments.  But I'm happy to support with reviews and general advice if someone is willing to drive this feature to completion.

**Comment by luisenp** [comment]
> Gotcha, so I'm guessing it's at the point where there are small issues and need to verify performance? @luisenp I just wanted to make sure there wasn't any more issues / blocking problems I missed skimming it.

There were some comments I left early that I'm not sure were addressed (mostly high level stuff). But leaving that aside I don't think the implementation was fully working yet, @Rohan138 would have more details though.

**Comment by natolambert** [comment]
Great. I want to take a look, and I have chatted with @danijar who didn't know it was being worked on. Let's see if I can un-stick it and if needed talk to Danijar.

**Comment by RajGhugare19** [comment]
Hello @natolambert -- I can take a lead developing this. You can review and sanity check the code afterwards. I will take a deeper look at the code and what changes are still required today.

**Comment by Rohan138** [comment]
Pitching in-I can help answer questions and debug the implementation over the weekend. The main function I'm unsure about is the DreamerAgent.train(...) function [here](https://github.com/facebookresearch/mbrl-lib/blob/71b54eb6097e6c2bf8b3d023ea3c98b8b49a3d31/mbrl/planning/dreamer_agent.py).

There's also some minor merge conflicts due to gym versions and gym's type checking that seem to be breaking CI; there's an open PR #161 here.
